dll/win32/msafd/misc/dllmain.c

```diff
@@ -3452,31 +3585,27 @@ SetSocketInformation(PSOCKET_INFORMATION Socket,
         if ((Socket->SharedData->CreateFlags & SO_SYNCHRONOUS_NONALERT) != 0)
         {
             TRACE("Opened without flag WSA_FLAG_OVERLAPPED. Do nothing.\n");
-            NtClose( SockEvent );
-            return 0;
-        }
-        if (CompletionRoutine == NULL)
-        {
-            /* Using Overlapped Structure, but no Completition Routine, so no need for APC */
-            APCContext = (PAFDAPCCONTEXT)Overlapped;
-            APCFunction = NULL;
-            Event = Overlapped->hEvent;
+            NtClose(SockEvent);
+            HeapFree(GlobalHeap, 0, InfoData);
+            return MsafdReturnWithErrno(STATUS_SUCCESS, NULL, 0, NULL);
         }
-        else
+
+        APCContext = HeapAlloc(GlobalHeap, 0, sizeof(AFDINFOAPCCONTEXT));
+        if (!APCContext)
         {
-            /* Using Overlapped Structure and a Completition Routine, so use an APC */
-            APCFunction = &AfdInfoAPC; // should be a private io completition function inside us
-            APCContext = HeapAlloc(GlobalHeap, 0, sizeof(AFDAPCCONTEXT));
-            if (!APCContext)
-            {
-                ERR("Not enough memory for APC Context\n");
-                NtClose( SockEvent );
-                return WSAEFAULT;
-            }
-            APCContext->lpCompletionRoutine = CompletionRoutine;
-            APCContext->lpOverlapped = Overlapped;
-            APCContext->lpSocket = Socket;
+            ERR("Not enough memory for APC Context\n");
+            NtClose(SockEvent);
+            HeapFree(GlobalHeap, 0, InfoData);
+            return MsafdReturnWithErrno(STATUS_INSUFFICIENT_RESOURCES, NULL, 0, NULL);
         }
+        APCContext->lpCompletionRoutine = CompletionRoutine;
+        APCContext->lpOverlapped = Overlapped;
+        APCContext->lpInfoData = InfoData;
+        APCContext->hSockEvent = SockEvent;
+        APCFunction = &AfdInfoAPC; // should be a private io completition function inside us
```

Comment:
```suggestion
        APCFunction = &AfdInfoAPC; // Should be a private IO completion function inside us
```
Now, what "inside us" may mean in this context, this is a mystery, but anyway.